src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -1055,16 +1055,71 @@ public void Connect(IPAddress[] addresses, int port)
                 throw new NotSupportedException(SR.net_invalidversion);
             }
 
+// Disable CS0162: Unreachable code detected
+//
+// SuportsMultipleConnectAttempts is a constant; when false, the following lines will trigger CS0162.
+#pragma warning disable 162
             Exception lastex = null;
-            foreach (IPAddress address in addresses)
+            if (SocketPal.SupportsMultipleConnectAttempts)
```

Comment:
+1 for @stephentoub's idea. (As I've mentioned in your previous PR, the pragma is the first thing that identifies a [code smell](https://en.wikipedia.org/wiki/Code_smell).)

Please add `//TODO`s throughout the code delimiting the new capability. I've noted that the new code introduces new states, etc which become very complicate to maintain in the future.

While I acknowledge that creating a PAL for Sockets is a huge task we should still aim towards that goal. For now, this PR is in my opinion technical dept and is very similar to code that we had before for Silverlight (that I've removed for CoreFX).